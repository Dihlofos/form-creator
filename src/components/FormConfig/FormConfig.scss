.form-config {
  &__button {
    margin: 20px 0 0;
  }
}
