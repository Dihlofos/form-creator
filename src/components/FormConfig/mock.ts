import { FormStructure } from 'store/types';

const FormMock: FormStructure = {
  title: 'Тестовая форма',
  action: '',
  method: 'POST',
  items: [
    {
      id: 'name',
      type: 'text',
      label: 'Имя',
      name: 'name',
      placeholder: 'Введите имя',
      defaultValue: 'Женя Самсонов',
    },
    {
      id: 'number',
      type: 'number',
      label: 'Числа',
      name: 'number',
      placeholder: 'Введите число',
      max: 10,
      min: 0,
      defaultValue: '5',
    },
    {
      id: 'textarea',
      type: 'textarea',
      label: 'Опишите нечто',
      name: 'textarea',
      placeholder: 'Введите многострочный текст',
      defaultValue: 'Многострочный текст',
    },
    {
      id: 'checkbox',
      type: 'checkbox',
      label: 'Чекбокс',
      name: 'checkbox',
      defaultChecked: true,
      placeholder: 'Согласны ли вы?',
    },
    {
      id: 'date',
      type: 'date',
      label: 'Дата',
      name: 'date',
      placeholder: 'Введите дату',
    },
    {
      title: 'Как вам в IT?',
      type: 'radio',
      name: 'howit',
      id: 'howit',
      radioValues: [
        {
          label: 'Отлично, нравится',
          value: 'great',
        },
        {
          label: 'Приемлемо',
          value: 'acceptable',
        },
        {
          label: 'Выгорел',
          value: 'burned',
        },
      ],
    },
  ],
  buttons: [
    {
      id: 'submit',
      label: 'Отправить',
      name: 'submit',
      action: 'FORM/SUBMIT',
    },
    {
      id: 'clear',
      label: 'Сбросить',
      name: 'clear',
      action: 'FORM/CLEAR',
    },
    {
      id: 'edit',
      label: 'Редактировать',
      name: 'edit',
      action: 'FORM/EDIT',
    },
  ],
};

export default FormMock;
