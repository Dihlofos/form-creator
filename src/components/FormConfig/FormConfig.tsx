import React, { ChangeEvent, FC, useCallback, useState } from 'react';
import { withTranslation, TFunction, WithTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';
import { setFormData } from 'store';
import { Button, TextArea } from 'ui';

import './FormConfig.scss';
import FormMock from './mock';

import FormJsonValidator from '../../utils/jsonValidation';

interface FormConfigProps extends WithTranslation {
  onConfigApply: () => void;
  t: TFunction;
}

const FormConfig: FC<FormConfigProps> = ({ t, onConfigApply }) => {
  const dispatch = useDispatch();
  const [config, setConfig] = useState<string>(
    JSON.stringify(FormMock, null, 2)
  );

  FormJsonValidator();

  const handleConfigChange = useCallback(
    (e: ChangeEvent<HTMLTextAreaElement>) => {
      const el = e.target as HTMLTextAreaElement;
      setConfig(el.value);
    },
    []
  );

  const handleApply = useCallback(() => {
    onConfigApply();
    dispatch(setFormData(JSON.parse(config)));
  }, [config, onConfigApply, dispatch]);

  return (
    <div className="form-config">
      <TextArea rows={30} value={config} onChange={handleConfigChange} />
      <Button className="form-config__button" onClick={handleApply}>
        {t('apply')}
      </Button>
    </div>
  );
};

export default withTranslation()(FormConfig);
