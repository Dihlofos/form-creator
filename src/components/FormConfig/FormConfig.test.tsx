import React from 'react';
import { create } from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import InitialState from '../../jest/InitialState';
import FormConfig from './FormConfig';

const mockStore = configureMockStore();

describe('Common App test', () => {
  it('FormTabs shows', () => {
    const store = mockStore(InitialState);

    const tree = create(
      <Provider store={store}>
        <FormConfig />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
