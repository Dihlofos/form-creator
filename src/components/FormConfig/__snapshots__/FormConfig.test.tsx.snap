// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common App test FormTabs shows 1`] = `
<div
  className="form-config"
>
  <div
    className="ant-row ant-form-item label-input-component"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        />
      </div>
    </div>
  </div>
  <textarea
    className="ant-input textarea"
    onChange={[Function]}
    onCompositionEnd={[Function]}
    onCompositionStart={[Function]}
    onKeyDown={[Function]}
    rows={30}
    style={Object {}}
    value="{
  \\"title\\": \\"Тестовая форма\\",
  \\"action\\": \\"\\",
  \\"method\\": \\"POST\\",
  \\"items\\": [
    {
      \\"id\\": \\"name\\",
      \\"type\\": \\"text\\",
      \\"label\\": \\"Имя\\",
      \\"name\\": \\"name\\",
      \\"placeholder\\": \\"Введите имя\\",
      \\"defaultValue\\": \\"Женя Самсонов\\"
    },
    {
      \\"id\\": \\"number\\",
      \\"type\\": \\"number\\",
      \\"label\\": \\"Числа\\",
      \\"name\\": \\"number\\",
      \\"placeholder\\": \\"Введите число\\",
      \\"max\\": 10,
      \\"min\\": 0,
      \\"defaultValue\\": \\"5\\"
    },
    {
      \\"id\\": \\"textarea\\",
      \\"type\\": \\"textarea\\",
      \\"label\\": \\"Опишите нечто\\",
      \\"name\\": \\"textarea\\",
      \\"placeholder\\": \\"Введите многострочный текст\\",
      \\"defaultValue\\": \\"Многострочный текст\\"
    },
    {
      \\"id\\": \\"checkbox\\",
      \\"type\\": \\"checkbox\\",
      \\"label\\": \\"Чекбокс\\",
      \\"name\\": \\"checkbox\\",
      \\"defaultChecked\\": true,
      \\"placeholder\\": \\"Согласны ли вы?\\"
    },
    {
      \\"id\\": \\"date\\",
      \\"type\\": \\"date\\",
      \\"label\\": \\"Дата\\",
      \\"name\\": \\"date\\",
      \\"placeholder\\": \\"Введите дату\\"
    },
    {
      \\"title\\": \\"Как вам в IT?\\",
      \\"type\\": \\"radio\\",
      \\"name\\": \\"howit\\",
      \\"id\\": \\"howit\\",
      \\"radioValues\\": [
        {
          \\"label\\": \\"Отлично, нравится\\",
          \\"value\\": \\"great\\"
        },
        {
          \\"label\\": \\"Приемлемо\\",
          \\"value\\": \\"acceptable\\"
        },
        {
          \\"label\\": \\"Выгорел\\",
          \\"value\\": \\"burned\\"
        }
      ]
    }
  ],
  \\"buttons\\": [
    {
      \\"id\\": \\"submit\\",
      \\"label\\": \\"Отправить\\",
      \\"name\\": \\"submit\\",
      \\"action\\": \\"FORM/SUBMIT\\"
    },
    {
      \\"id\\": \\"clear\\",
      \\"label\\": \\"Сбросить\\",
      \\"name\\": \\"clear\\",
      \\"action\\": \\"FORM/CLEAR\\"
    },
    {
      \\"id\\": \\"edit\\",
      \\"label\\": \\"Редактировать\\",
      \\"name\\": \\"edit\\",
      \\"action\\": \\"FORM/EDIT\\"
    }
  ]
}"
  />
  <div
    className="form-config__bottom"
  >
    <button
      className="ant-btn form-config__button"
      onClick={[Function]}
      type="button"
    />
  </div>
</div>
`;
